File: src/domain_ts/Palette.ts
import { Color, ColorSpace } from "./Color";

type DivisionPoints = number[];
type Colors = Array<Color>;

export enum Themes {
  GRAY_SCALE,
  CYCLIC_GRAY_SCALE,
  RED_CYAN,
  BLUE_GOLD,
  EARTH_AND_SKY,
  HOT_AND_COLD,
  FIRE,
  TREE_COLORS,
  SEA_SHORE,
  RANDOM
}

const THEMES_MAP: {
  [key: string]: () => {
    divisionPoints: DivisionPoints;
    colors: Colors;
  };
} = {
  [Themes.GRAY_SCALE]: () => ({
    divisionPoints: [0, 1],
    colors: [
      new Color(1, 1, 1),
      new Color(0, 0, 0),
    ]
  }),
  [Themes.CYCLIC_GRAY_SCALE]: () => ({
    divisionPoints: [0, 0.5, 1],
    colors: [
      new Color(0, 0, 0),
      new Color(1, 1, 1),
      new Color(0, 0, 0),
    ]
  }),
  [Themes.RED_CYAN]: () => ({
    divisionPoints: [0, 0.5, 1],
    colors: [
      new Color(1, 0, 0),
      new Color(0, 1, 1),
      new Color(1, 0, 0),
    ]
  }),
  [Themes.BLUE_GOLD]: () => ({
    divisionPoints: [0, 0.5, 1],
    colors: [
      new Color(0.1, 0.1, 1),
      new Color(1, 0.6, 0),
      new Color(0.3, 0.3, 1),
    ]
  }),
  [Themes.EARTH_AND_SKY]: () => ({
    divisionPoints: [0, 0.15, 0.33, 0.67, 0.85, 1],
    colors: [
      new Color(1, 1, 1),
      new Color(1, 0.8, 0),
      new Color(0.53, 0.12, 0.075),
      new Color(0, 0, 0.6),
      new Color(0, 0.4, 1),
      new Color(1, 1, 1),
    ]
  }),
  [Themes.HOT_AND_COLD]: () => ({
    divisionPoints: [0, 0.16, 0.5, 0.84, 1],
    colors: [
      new Color(1, 1, 1),
      new Color(0, 0.4, 1),
      new Color(0.2, 0.2, 0.2),
      new Color(1, 0, 0.8),
      new Color(1, 1, 1),
    ]
  }),
  [Themes.FIRE]: () => ({
    divisionPoints: [0, 0.17, 0.83, 1],
    colors: [
      new Color(0, 0, 0),
      new Color(1, 0, 0),
      new Color(1, 1, 0),
      new Color(1, 1, 1),
    ]
  }),
  [Themes.SEA_SHORE]: () => ({
    divisionPoints: [0, 0.1667, 0.3333, 0.5, 0.6667, 0.8333, 1],
    colors: [
      new Color(0.7909, 0.9961, 0.763),
      new Color(0.8974, 0.8953, 0.6565),
      new Color(0.9465, 0.3161, 0.1267),
      new Color(0.5184, 0.1109, 0.0917),
      new Color(0.0198, 0.4563, 0.6839),
      new Color(0.5385, 0.8259, 0.8177),
      new Color(0.7909, 0.9961, 0.76)
    ]
  })
};

/**
 * Palette & themes originally made by David Eck
 * http://math.hws.edu/eck/js/mandelbrot/MB.html
 */
export class Palette {
  constructor(
    private divisionPoints: DivisionPoints,
    private divisionColors: Colors
  ) {}

  toColor (color: Color | number[]): Color {
    if (color instanceof Color) {
      return color
    }
    return new Color(color[0], color[1], color[2])
  }

  getNearestDivisionPoint (position: number) {
    let nearestDivisionPoint = 1;
    while (position > this.divisionPoints[nearestDivisionPoint]) {
      nearestDivisionPoint++;
    }
    return nearestDivisionPoint
  }

  /**
   * Resturns the Color that corresponds to the given Position
   * @param position Value between 0 and 1
   */
  public getColor(position: number): Color {
    const nearestDivisionPoint = this.getNearestDivisionPoint(position)
    const p1 = this.divisionPoints[nearestDivisionPoint - 1]
    const p2 = this.divisionPoints[nearestDivisionPoint]

    const ratio = (position - p1) / (p2 - p1);

    return Color.colorBetween(
      this.divisionColors[nearestDivisionPoint - 1],
      this.divisionColors[nearestDivisionPoint],
      ratio
    )
  }

  static createStandardPalette(theme: Themes): Palette {
    const { divisionPoints, colors } = THEMES_MAP[theme]();
    return new Palette(divisionPoints, colors);
  }
}

File: src/MandelbrotSet/infrastructure/MandelbrotSetWasm/rust/mandelbrot_set/palette.rs
mod color;

use self::color::Color;

type DivisionPoints = Vec<f64>;
type DivisionColors = Vec<Color>;

pub enum Theme {
  GrayScale,
  CyclicGrayScale,
  RedCyan,
  BlueGold,
  EarthAndSky,
  HotAndCold,
  Fire,
  TreeColors,
  SeaShore,
}

pub struct Palette {
  division_points: DivisionPoints,
  division_colors: DivisionColors,
}

impl Palette {
  pub fn new(division_points: DivisionPoints, division_colors: DivisionColors) -> Palette {
    Palette {
      division_points,
      division_colors,
    }
  }

  pub fn get_nearest_division_point(&self, position: f64) -> usize {
    let mut nearest_division_point = 1;
    while position > self.division_points[nearest_division_point] {
      nearest_division_point = nearest_division_point + 1;
    }
    return nearest_division_point;
  }

  /**
   * Resturns the Color that corresponds to the given Position
   * @param position Value between 0 and 1
   */
  pub fn get_color(&self, position: f64) -> Color {
    let nearest_division_index = self.get_nearest_division_point(position);
    let left_point = self.division_points[nearest_division_index - 1];
    let right_point = self.division_points[nearest_division_index];

    let ratio = (position - left_point) / (right_point - left_point);

    return Color::color_between(
      &self.division_colors[nearest_division_index - 1],
      &self.division_colors[nearest_division_index],
      ratio,
    );
  }

  pub fn create_standard_palette(theme: Theme) -> Palette {
    match theme {
      Theme::GrayScale => Palette::new(
        vec![0.0, 0.1],
        vec![
          Color::new(1.0, 1.0, 1.0),
          Color::new(0.0, 0.0, 0.0),
        ],
      ),
      Theme::CyclicGrayScale => Palette::new(
        vec![0.0, 0.1],
        vec![
          Color::new(1.0, 1.0, 1.0),
          Color::new(0.0, 0.0, 0.0),
        ],
      ),
      Theme::RedCyan => Palette::new(
        vec![0.0, 0.1],
        vec![
          Color::new(1.0, 1.0, 1.0),
          Color::new(0.0, 0.0, 0.0),
        ],
      ),
      Theme::BlueGold => Palette::new(
        vec![0.0, 0.1],
        vec![
          Color::new(1.0, 1.0, 1.0),
          Color::new(0.0, 0.0, 0.0),
        ],
      ),
      Theme::EarthAndSky => Palette::new(
        vec![0.0, 0.15, 0.33, 0.67, 0.85, 1.0],
        vec![
          Color::new(1.0, 1.0, 1.0),
          Color::new(1.0, 0.8, 0.0),
          Color::new(0.53, 0.12, 0.075),
          Color::new(0.0, 0.0, 0.6),
          Color::new(0.0, 0.4, 1.0),
          Color::new(1.0, 1.0, 1.0),
        ],
      ),
      Theme::HotAndCold => Palette::new(
        vec![0.0, 0.1],
        vec![
          Color::new(1.0, 1.0, 1.0),
          Color::new(0.0, 0.0, 0.0),
        ],
      ),
      Theme::Fire => Palette::new(
        vec![0.0, 0.17, 0.83, 1.0],
        vec![
          Color::new(0.0, 0.0, 0.0),
          Color::new(1.0, 0.0, 0.0),
          Color::new(1.0, 1.0, 0.0),
          Color::new(1.0, 1.0, 1.0),
        ],
      ),
      Theme::TreeColors => Palette::new(
        vec![0.0, 0.1],
        vec![
          Color::new(1.0, 1.0, 1.0),
          Color::new(0.0, 0.0, 0.0),
        ],
      ),
      Theme::SeaShore => Palette::new(
        vec![0.0, 0.1],
        vec![
          Color::new(1.0, 1.0, 1.0),
          Color::new(0.0, 0.0, 0.0),
        ],
      ),
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn it_works() {
    let palette = Palette::create_standard_palette(Theme::EarthAndSky);
    let color = palette.get_color(0.1);

    assert_eq!(color.get_r(), 255);
    assert_eq!(color.get_g(), 221);
    assert_eq!(color.get_b(), 85);
  }

  #[test]
  fn it_works_two() {
    let palette = Palette::create_standard_palette(Theme::EarthAndSky);
    let color = palette.get_color(0.2);

    assert_eq!(color.get_r(), 222);
    assert_eq!(color.get_g(), 156);
    assert_eq!(color.get_b(), 5);
  }
}